File: string_problems/SortCharacterByFrequency.cpp

// Problem link - https://leetcode.com/problems/sort-characters-by-frequency/?envType=problem-list-v2&envId=string

#include<bits/stdc++.h>
using namespace std;

string frequencySort(string s) {
    // step 1 - use map to store frequency of each character
    unordered_map<char, int> frqMap;
    for(char ch : s){
        frqMap[ch]++;
    }

    // step 2 - extract characters into vector
    vector<char> chars;
    for(const auto& pair : frqMap){
        chars.push_back(pair.first);
    }

    // step 3 - sort character by its frequency
    sort(chars.begin(), chars.end(), [&frqMap](char a, char b){
        // sorting in desceding order
        return frqMap[b] < frqMap[a];
    });

    // step 4 - build the result
    string res;
    for(char ch : chars){
        res += string(frqMap[ch], ch);
    }

    // step 5 - return resultant string `res`
    return res;
}

int main(){
    // string s = "tree";

    // string s = "cccaaa";

    string s = "Aabb";
    cout<<frequencySort(s)<<endl;
    return 0;
}

File: string_problems/SortCharacterByFrequency.js

// Problem link - https://leetcode.com/problems/sort-characters-by-frequency/?envType=problem-list-v2&envId=string

let frequencySort = function(s) {
    // step 1 - store frquency of each character in frquency map `frqMap` object
    let frqMap = {};
    for(let ch of s){
        frqMap[ch] = (frqMap[ch] || 0) + 1;
    }

    // return result 
    return Object.keys(frqMap)
        // step 2 - sort character by its frequency in descending order  
        .sort((a,b) => frqMap[b] - frqMap[a])
        // step 3 - repeat each character by its frequency
        .map((char) => char.repeat(frqMap[char]))
        // step 4 - join back to string and return
        .join('');
};

// const s = "tree";

// const s = "cccaaa";

// const s = "Aabb";

const s = "loveleetcode";
console.log(frequencySort(s));
